crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+use std::iter::FusedIterator;
+
+use ruff_db::file_system::{FileSystemPath, FileSystemPathBuf};
+use ruff_db::vfs::{system_path_to_file, VfsPath};
+
+use crate::module_name::ModuleName;
+use crate::supported_py_version::get_target_py_version;
+use crate::typeshed::{parse_typeshed_versions, TypeshedVersionsQueryResult};
+use crate::{Db, TypeshedVersions};
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct ExtraPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct ExtraPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct FirstPartyPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct FirstPartyPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct StandardLibraryPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct StandardLibraryPathBuf(FileSystemPathBuf);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub(crate) struct SitePackagesPath(FileSystemPath);
+
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub(crate) struct SitePackagesPathBuf(FileSystemPathBuf);
```

Comment:
To me the benefit of having all these different `Path` variants isn't entirely clear. I get the impression that we now use two forms to distinguish the paths:

* The variant in `ModuleResolutionPath` 
* The different path types

The only place where I see that we pass a specific path to a method is `StandardLibraryPath`. Considering that this method is private, I think it's fine to just pass `FileSystemPath`. 

I would recommend removing the different path versions and just use `FileSystemPathBuf` everywhere and solely rely on the enum variant to distinguish the different cases.